nemo/collections/asr/models/wav2vec/wav2vec_model.py

```diff
@@ -0,0 +1,541 @@
+        cfg = OmegaConf.create(OmegaConf.to_container(cfg, resolve=True))
+        cfg = OmegaConf.merge(schema, cfg)
+
+        feature_enc_layers = cfg.conv_feature_encoder.conv_feature_layers
+        self.embed = feature_enc_layers[-1][0]  # Select last conv output layer dimension
+
+        self.feature_extractor = ConvFeatureEncoder(
+            conv_layers=feature_enc_layers,
+            mode=cfg.conv_feature_encoder.extractor_mode,
+            conv_bias=cfg.conv_feature_encoder.conv_bias,
+        )
+
+        encoder_embed_dim = cfg.transformer_encoder.encoder.embedding_dim
+        self.post_extract_proj = (
+            nn.Linear(self.embed, encoder_embed_dim)
+            if self.embed != encoder_embed_dim and not cfg.quantize.quantize_input
+            else None
+        )
+
+        self.mask_cfg = cfg.mask
+
+        self.dropout_input = nn.Dropout(cfg.dropout_input)
+        self.dropout_features = nn.Dropout(cfg.dropout_features)
+
+        self.feature_grad_mult = cfg.feature_grad_mult
+
+        self.quantizer = None
+        self.input_quantizer = None
+
+        self.n_negatives = cfg.n_negatives
+        self.cross_sample_negatives = cfg.cross_sample_negatives
+        self.codebook_negatives = cfg.codebook_negatives
+        self.negatives_from_everywhere = cfg.negatives_from_everywhere
+
+        final_dim = cfg.final_dim if cfg.final_dim > 0 else encoder_embed_dim
+        self.final_dim = final_dim
+        if cfg.quantize.quantize_targets:
+            vq_dim = cfg.quantize.latent_dim if cfg.quantize.latent_dim > 0 else final_dim
+            self.quantizer = GumbelVectorQuantizer(
+                dim=self.embed,
+                num_vars=cfg.quantize.latent_vars,
+                temp=cfg.quantize.latent_temp,
+                groups=cfg.quantize.latent_groups,
+                combine_groups=False,
+                vq_dim=vq_dim,
+                time_first=True,
+            )
+            self.project_q = nn.Linear(vq_dim, final_dim)
+        else:
+            self.project_q = nn.Linear(self.embed, final_dim)
+
+        if cfg.quantize.quantize_input:
+            if cfg.quantize.same_quantizer and self.quantizer is not None:
+                vq_dim = final_dim
+                self.input_quantizer = self.quantizer
+            else:
+                vq_dim = cfg.quantize.latent_dim if cfg.quantize.latent_dim > 0 else encoder_embed_dim
+                self.input_quantizer = GumbelVectorQuantizer(
+                    dim=self.embed,
+                    num_vars=cfg.quantize.latent_vars,
+                    temp=cfg.quantize.latent_temp,
+                    groups=cfg.quantize.latent_groups,
+                    combine_groups=False,
+                    vq_dim=vq_dim,
+                    time_first=True,
+                )
+            self.project_inp = nn.Linear(vq_dim, encoder_embed_dim)
+
+        self.mask_emb = nn.Parameter(torch.FloatTensor(encoder_embed_dim).uniform_())
```

Comment:
Is mask_emb learnable?

Reply:
This was ported from the fairseq code, given it's a parameter I'm going to assume so.